### ISSUE TITLE
Using completed tasks as release date for 3.0

### ISSUE THREAD

**Issue by jonaslagoni** [asyncapi/spec#944]
In https://github.com/asyncapi/spec/issues/880 we tried to figure out a specific release date, however, instead of using dates, I am proposing that we use tasks as the finish goal when we should release 3.0. So once all those tasks are complete, we can release it in whatever spec release period we are in. Of course, we can make guesses on when that will happen, but tasks talk.

The reason I am proposing this is because it's open source 🤷 Contributors have different priorities and things just take longer than expected, and sticking to a schedule in a non-paying (for most) environment just is not feasible or even to be expected 😄 

So here are the remaining issues I propose all should be completed before release:

Specification:
- [x] https://github.com/asyncapi/spec/issues/505
- [x] https://github.com/asyncapi/spec/issues/622

For tooling this is what I think is the bare minimum needed to work with AsyncAPI:

- [x] https://github.com/asyncapi/spec-json-schemas/issues/400
- [x] https://github.com/asyncapi/parser-js/issues/780
- [x] https://github.com/asyncapi/generator/issues/979

For docs:
- [ ] Migration guides: https://github.com/asyncapi/website/pull/2008

Of course, this does NOT mean that other libraries should not be updated or worked on in the meantime, this should just be the bare minimum that blocks the release of 3.0.

**Comment by fmvilas** [comment]
Yeah, agree! We should use this as the "roadmap" for when to release it instead of dates.

**Comment by jonaslagoni** [comment]
After https://github.com/asyncapi/community/issues/737 it seems https://github.com/asyncapi/server-api/issues/294 needs to be added to the list because of the studio and CLI usages.

**Comment by Amzani** [comment]
Using completion of tasks as a criterion for the release of version 3.0 of the project sounds like a reasonable approach, considering the challenges of coordinating contributors and the unpredictable nature of open-source development!

We could use https://shapeit.app/ to make this visible and make it fun

**Comment by jonaslagoni** [comment]
I am suggesting that we cut our requirements way down to release v3 to just the very core that will enable others to update in due time.

This means that we are effectively only relying on https://github.com/asyncapi/generator/issues/979 to be complete before we release v3, which already is almost done: https://github.com/asyncapi/generator/pull/1017

For docs that means the only thing that is missing is merging https://github.com/asyncapi/website/pull/2008.

What ever other tool that is updated at the time of the release, perfect, if not, also perfect. It will happen in due time, but at least it is not blocking v3 release.

This decision takes effect Monday September 11th using lazy consensus, unless a codeowner of the spec objects. This also means that I aim the v3 release for the last week of September.

**Comment by fmvilas** [comment]
I agree with your statement, @jonaslagoni. I'm happy to consider adding the react component as a blocker since it's very close to being done. What do you folks think?

**Comment by derberg** [comment]
- React-component -> Studio and HTML template
- Basic docs migrated (the ones that we have for 2.x already)

☝🏼 these are must have for me

Docs generation is main use case for AsyncAPI. Even if company uses AsyncAPI for other things like code generation, broker management and other stuff - docs generation is another "easy to get" side effect.

---

What if we make an agreement that if there will be no volunteers to commit to put priority on given task, then it will not be taken into the scope?

**Comment by jonaslagoni** [comment]
> What if we make an agreement that if there will be no volunteers to commit to put priority on given task, then it will not be taken into the scope?

Then it's the same as my proposal, just with a longer fuse.

If it makes it in time, it makes it, otherwise it won't 🤷

I just don't want it to be blocking the release.

**Comment by jonaslagoni** [comment]
Yeah, the problem is scope and commitment keep changing and including stuff left and right. By minimizing the scope to the bare minimum as suggested above, we already know the concrete stuff that is needed.

To me, all the other features are just nice to have.

Let me just say it plainly, if we want to include anything extra than what I suggested above, we will need to find another person to coordinate that effort 🙂 

I will of course finish off what is listed above. Once you all reach an agreement on what needs to change for us to release v3 (docs, code, and what not), and they are done, then I can step in again and help finish the last tasks as release coordinator to release v3. 

We are already way above and beyond what I expected was needed as a release coordinator so this is where we need someone else to step up 🙂

**Comment by derberg** [comment]
No worries, I think you made it pretty clear during our last public call. What are you plans for next public call? since we travel to AsyncAPI conference next week

**Comment by fmvilas** [comment]
About the next public call, we can have it the week after.

> What if we make an agreement that if there will be no volunteers to commit to put priority on given task, then it will not be taken into the scope?

This is super important IMHO. Let's try to make it happen but if we can't, we ship with whatever we have. I agree with this.

Also, folks, remember that the release date doesn't have to be the same as the communication/promotion/announcement date. If we see that some things are not done but they're close to being done, we can postpone announcing it. This way at least we get rid of the release duty.

**Comment by smoya** [comment]
> What if we make an agreement that if there will be no volunteers to commit to put priority on given task, then it will not be taken into the scope?

When we talk about volunteers, aren't Code Owners of the repositories we talk about needed to be involved? I.e. Code Owners from https://github.com/asyncapi/asyncapi-react.
I mean we would need them in order to finally decide if they want to commit to it, even though they do not code the changes.

**Comment by jonaslagoni** [comment]
Mad release coordinator here 👼 

With all the tasks being done that I proposed, and ontop of that react components, HTML template, markdown template, studio, generator, CLI, and parser all supporting v3, I am going to start the release process for v3 on **the 25th of October**.

With CI run times this probably means that the release will be done on the **27th**.

We have achieved all the steps on the release process and then some. So unless a spec code owner vetoes this decision (@fmvilas @derberg @smoya @dalelane @char0n, vetoing also means replacing me as release coordinator with yourself), that is what I am gonna aim for.

Regarding outstanding website and docs changes, I simply don't care, migration guide has been standing still for nearly two months, and so for the other changes. Either the website is ready for when the spec is released or it's not.

The same goes for any other small issues, they are either gonna be fixed at the time of the release, or it's gonna happen afterwards, but I see nothing else blocking the release process.

**Comment by fmvilas** [comment]
I'm all for it. Let's release it on the 25th. By then we may even get Glee supporting it too.

**Comment by smoya** [comment]
As I mentioned in [this Slack thread](https://asyncapi.slack.com/archives/C0230UAM6R3/p1697451314230319?thread_ts=1697448403.662059&cid=C0230UAM6R3), I want to clarify people that [V3 Spec reference page](https://www.asyncapi.com/docs/reference/specification/v3.0.0) will be out by then. 

You have my 👍. Happy to help with any required stuff.

**Comment by derberg** [comment]
For me there are certain "basics that need to be done" or "unknowns" that block the release:

- Documentation generation:
  - React component fully supporting AsyncAPI v3 -> https://github.com/asyncapi/asyncapi-react/pull/795 (request/reply)
  - No more PUB/SUB but RECEIVES/SENDS
  - Making sure latest React is bumped in HTML template and Studio
- Studio updates:
  - We should suggest v3 conversion for old documents (the modal that suggest conversion I mean) as now we suggest conversion from 3.0 to 2.6 (which will even not work)
  - Making sure latest parser is used
  - We need updates examples in Studio + also example of v3 should load on first interaction
- Spec issues:
  - We need clear stand from spec maintainers on regression in functionality we provide with params -> https://github.com/asyncapi/spec/issues/979
  - I'm very unsure about "messageId" as it looks like a left over that will only be confusing in future and reason for removal and breaking changes -> https://github.com/asyncapi/spec/issues/979
  - Spec examples are misleading -> https://github.com/asyncapi/spec/pull/977
- Docs:
  - We need to setup v2 docs -> https://github.com/asyncapi/website/issues/1650
  - These docs need an update, except for tutorial that unfortunately will have to be removed until related template is migrated -> https://github.com/asyncapi/website/issues/created_by/jonaslagoni
  - Making sure acknowledgement/migration/notes are merged and ready
- Parser?
  - A confusion starts because of Pub/Sub and I think it is related to parser -> https://github.com/asyncapi/asyncapi-react/issues/815. We need a clear answer what is the reason. How is this related to v3? as this confusion affects latest React component. We basically need a way for docs generation to say that prior 3.0 you get in the UI PUB/SUB the right way and starting from 3.0 you get RECIVES/SEND the right way. 

---

As spec code owner I cannot imagine giving 🟢 light to release, if the next day someone comes to me and asks about, for example, `messageId` and I do not know the answer. 

Some of the things from above list (development related) can be done few days after release. For me critical is to know a commitment. Who can commit to resolve certain items as a priority.

So, yeah, I already wrote in https://github.com/asyncapi/spec/issues/944#issuecomment-1714091232 that setting any date without knowing if anyone commits to help is not going to work. We should really not follow the practice that we just set a date, and not care if people will make it or not. 

I'm happy to help figuring out commitments and help pushing for answers.

**Comment by derberg** [comment]
there was a wrong link to `messageId` topic, updated to https://github.com/asyncapi/spec/issues/978

**Comment by Amzani** [comment]
Studio updates:

>We should suggest v3 conversion for old documents (the modal that suggest conversion I mean) as now we suggest conversion from 3.0 to 2.6 (which will even not work)

https://github.com/asyncapi/studio/issues/832

> Making sure latest parser is used

✅ 

> We need updates examples in Studio + also example of v3 should load on first interaction

https://github.com/asyncapi/studio/issues/831

**Comment by derberg** [comment]
ok, so I got confirmation from @AceTheCreator that he will help with **Documentation generation** topic I mentioned and @alequetzalli will be there active helping to review and approve stuff I will be working on in **Docs** topic. So for this topics I'm pretty sure we will be ready by 15th November. 

Also @jonaslagoni you have my full support with a release itself. We can do a fun live stream with beers and crackers, or champaign and strawberries, up to you 😆

**Comment by fmvilas** [comment]
That's great. Thanks for coming up with a date, @derberg. Let me also coordinate with @thulieblack for the announcement and marketing activities. I propose we **release and announce** v3 in December so we have time to prepare all the marketing material and plan stuff accordingly. Does that sound good, @derberg?

**Comment by jonaslagoni** [comment]
Thanks @derberg, that will definitely help make the progress specific as to what is needed 👍 Thanks @Amzani for creating the studio outstanding tasks 👌 

So here is the cleaned-up task list (what is missing and in progress), do you mind double-checking @derberg to ensure I got everything? I have included everything BUT docs as I read it like you push that effort with help from @AceTheCreator and @alequetzalli. So the below list is what I will help bring awareness to 🙂 

New tasks (not in progress):
- [ ] https://github.com/asyncapi/asyncapi-react/issues/815 - No more PUB/SUB but RECEIVES/SENDS
- [ ] https://github.com/asyncapi/studio/issues/832 - We should suggest v3 conversion for old documents (the modal that suggest conversion I mean) as now we suggest conversion from 3.0 to 2.6 (which will even not work)
- [ ] https://github.com/asyncapi/studio/issues/831 - We need updates examples in Studio + also example of v3 should load on first interaction
- [ ] https://github.com/asyncapi/spec/issues/979 - We need clear stand from spec maintainers on regression in functionality we provide with params 
- [ ] https://github.com/asyncapi/spec/issues/978 - I'm very unsure about "messageId" as it looks like a left over that will only be confusing in future and reason for removal and breaking changes


Tracking list (tasks list already in progress)
- [ ] https://github.com/asyncapi/asyncapi-react/pull/795 - React component fully supporting AsyncAPI v3 which means request/reply rendering.
- [ ] https://github.com/asyncapi/spec/pull/977 - Spec examples are misleading

> Making sure latest React is bumped in HTML template and Studio

This is already automatic, we just have to ensure the pipeline don't fail etc, but no manual work should be necessary.

> A confusion starts because of Pub/Sub and I think it is related to parser -> https://github.com/asyncapi/asyncapi-react/issues/815. We need a clear answer what is the reason. How is this related to v3? as this confusion affects latest React component. We basically need a way for docs generation to say that prior 3.0 you get in the UI PUB/SUB the right way and starting from 3.0 you get RECIVES/SEND the right way.

This is just a problem with react components, parser explicitly tells what the operation is, if you want to revert it for v2 (as we do in react components), then it needs to happen in each individual tool. Linked it as a task in the list above https://github.com/asyncapi/asyncapi-react/issues/815

> Let me also coordinate with @thulieblack for the announcement and marketing activities. I propose we release and announce v3 in December so we have time to prepare all the marketing material and plan stuff accordingly. Does that sound good, @derberg?

👌 @fmvilas would it be possible for you to come up with a list of tasks and commitments as @derberg just did for the above activities?

**Comment by fmvilas** [comment]
> 👌 @fmvilas would it be possible for you to come up with a list of tasks and commitments as @derberg just did for the above activities?

Yes, on it. I'm gonna put together a suggestion for a marketing plan and discuss it with @thulieblack. Once we agree on how to proceed, I'll create tasks for it. I'll put a link to the plan here once it's done.

**Comment by derberg** [comment]
> A confusion starts because of Pub/Sub and I think it is related to parser -> https://github.com/asyncapi/asyncapi-react/issues/815. We need a clear answer what is the reason. How is this related to v3? as this confusion affects latest React component. We basically need a way for docs generation to say that prior 3.0 you get in the UI PUB/SUB the right way and starting from 3.0 you get RECIVES/SEND the right way.

can confirm ☝🏼 was a bug in react component and below PR will fix it

https://github.com/asyncapi/asyncapi-react/pull/795

**Comment by derberg** [comment]
just marked below as solved

![Screenshot 2023-11-16 at 17 45 59](https://github.com/asyncapi/spec/assets/6995927/875253c0-3e2f-4393-983a-6a6f4fbd8290)

also majority of docs are in v3 subdomain - only missing PR is https://github.com/asyncapi/website/pull/2317

my only concerns is:
- during req/reply development in react-component, few more bugs with parser were found
- new confusion after last 2 weeks of focus on v3 -> https://github.com/asyncapi/spec/issues/991

@Amzani you can work on studio update to use latest react component
